include/seastar/core/abort_source.hh

```diff
@@ -96,33 +103,45 @@ public:
 private:
     using subscription_list_type = bi::list<subscription, bi::constant_time_size<false>>;
     std::optional<subscription_list_type> _subscriptions = subscription_list_type();
+    std::exception_ptr _ex = nullptr;
 
 public:
     /// Delays the invocation of the callback \c f until \ref request_abort() is called.
     /// \returns an engaged \ref optimized_optional containing a \ref subscription that can be used to control
     ///          the lifetime of the callback \c f, if \ref abort_requested() is \c false. Otherwise,
     ///          returns a disengaged \ref optimized_optional.
+    template <typename Func>
     [[nodiscard]]
-    optimized_optional<subscription> subscribe(subscription_callback_type f) noexcept(std::is_nothrow_move_constructible<subscription_callback_type>::value) {
+    optimized_optional<subscription> subscribe(Func&& f) noexcept(std::is_nothrow_move_constructible_v<Func>) {
         if (abort_requested()) {
             return { };
         }
-        return { subscription(*this, std::move(f)) };
+        if constexpr (std::is_invocable_v<Func, std::exception_ptr>) {
+            return { subscription(*this, std::forward<Func>(f)) };
+        } else {
+            return { subscription(*this, std::forward<Func>(f), subscription::naive_cb_tag{}) };
+        }
     }
 
     /// Requests that the target operation be aborted. Current subscriptions
     /// are invoked inline with this call, and no new ones can be registered.
     /// Must be called exactly once, otherwise the program will be aborted.
-    void request_abort() noexcept {
+    void request_abort(std::exception_ptr ex = std::make_exception_ptr(abort_requested_exception())) noexcept {
         assert(_subscriptions);
+        _ex = std::move(ex);
         auto subs = std::exchange(_subscriptions, std::nullopt);
         while (!subs->empty()) {
             subscription& s = subs->front();
             s.unlink();
-            s.on_abort();
+            s.on_abort(_ex);
```

Comment:
I do like the patch, but after thinking about how it (sh)would proceed next have some concerns. Let's consider the `sleep_abortable(duration, abort_source&)` thing. Internally it looks like

```c++
    ...
        sleeper(typename Clock::duration dur, abort_source& as)
                : tmr([this] { done.set_value(); }) {
            auto sc_opt = as.subscribe([this] () noexcept {
                if (tmr.cancel()) {
                    done.set_exception(sleep_aborted());
                }
            });
    ...
```

so right now `as.request_abort()` call resolves this into `sleep_aborted`.

If patching this code to support exception propagation from `as.request_abort(exception)` the code would probably look like below

```c++
    ...
        sleeper(typename Clock::duration dur, abort_source& as)
                : tmr([this] { done.set_value(); }) {
            auto sc_opt = as.subscribe([this] (std::exception_ptr ex) noexcept {
                if (tmr.cancel()) {
                    done.set_exception(ex);
                }
            });
    ...
```

so that `as.request_abort(ex)` would correctly propagate `ex` up the waiting chain and resolve sleep future into `ex`. However, the legacy code that does `as.request_abort()` would set `ex` to default `abort_requested_exception` and it will thus silently replace the former `sleep_aborted` one.

Reply:
Added a virtual `get_default_exception()` method provided for customized abort sources (by inheritance and override)